src/redprl/refiner.fun

```diff
@@ -1015,8 +1013,190 @@ struct
+      if P.eq Sym.eq (ext, eps) then [jdg] (* combining rules in first row *)
+      else if neqDims (ext, eps) then [] (* second row, left rule *)
+      else
+        let
+          val (P.VAR v, P.APP eps) = (ext, eps)
+        in
+          [Seq.map (substSymbol (P.APP eps, v)) jdg] (* bottom left rule *)
+        end
+
+    (* Restrict a judgement by two equations `ext0 = eps0` and `ext1 = eps1`. *)
+    fun Two (jdg : abt jdg) (ext0 : param) (eps0 : param) (ext1 : param) (eps1 : param) =
+      if P.eq Sym.eq (ext0, eps0) then One jdg ext1 eps1 (* top right rule *)
+      else if P.eq Sym.eq (ext1, eps1) then One jdg ext0 eps0 (* top right rule *)
+      else if neqDims (ext0, eps0) then [] (* second row, left rule *)
+      else if neqDims (ext1, eps1) then [] (* second row, left rule *)
+      else
+        let
+          val (P.VAR u, P.APP _) = (ext0, eps0)
+          val (P.VAR v, P.APP _) = (ext1, eps1)
+        in
+          if Sym.eq (u, v) andalso not (P.eq Sym.eq (eps0, eps1)) then [] (* second row, right rule *)
+          else
+            [Seq.map (substSymbol (eps0, u)) (Seq.map (substSymbol (eps1, v)) jdg)] (* bottom right rule *)
+        end
+  end
+
+  structure HCom =
+  struct
+    (* Given a list of n extents and a list of 2n tubes, produce a list of
+       (extent * side * tube) triples, where side is 0 or 1 according to the
+       tube's position in the list *)
+    fun groupTubes [] [] = []
+      | groupTubes (ext :: exts) (tube0 :: tube1 :: tubes)  =
+        (ext, P.APP P.DIM0, tube0) :: (ext, P.APP P.DIM1, tube1) :: groupTubes exts tubes
+      | groupTubes _ _ = raise Fail "groupTubes"
+
+    fun listToTel l = List.foldl (fn (g, l) => l >: g) T.empty l
+
+    (* Produce the list of goals requiring that tube aspects agree with each other. *)
+    fun intraTubeGoals H w ty group =
+      let
+        fun intraTube ext0 eps0 (u, tube0) ext1 eps1 (v, tube1) =
+          let
+            val tube0 = substSymbol (P.ret w, u) tube0
+            val tube1 = substSymbol (P.ret w, v) tube1
+            val J = H >> CJ.EQ ((tube0,tube1), ty)
+          in
+            List.map (fn j => #1 (makeGoal (([(w, P.DIM)], []) |> j)))
+                     (Restriction.Two J ext0 eps0 ext1 eps1)
+          end
+      in
+        listToTel
+          (ListMonad.bind (fn (ext0, eps0, tube0) =>
+             ListMonad.bind (fn (ext1, eps1, tube1) =>
+                 intraTube ext0 eps0 tube0 ext1 eps1 tube1)
+               group)
+             group)
+      end
+
+    (* Produce the list of goals requiring that tube aspects agree with the cap. *)
+    fun tubeCapGoals H ty r cap group =
+      let
+        fun tubeCap ext eps (u, tube) =
+          let
+            val J = H >> CJ.EQ ((substSymbol (r,u) tube, cap), ty)
+          in
+            List.map (fn j => #1 (makeGoal j)) (Restriction.One J ext eps)
+          end
+      in
+        listToTel (ListMonad.bind (fn (ext, eps, tube) => tubeCap ext eps tube) group)
```

Comment:
If you change `tubeCap` to be curried, this could be written `listToTel (ListMonad.find tubeCap group)`, which might be a little cleaner.